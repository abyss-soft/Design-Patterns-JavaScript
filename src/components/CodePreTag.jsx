import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';

const CodePreTag = ({ children, ...restProps }) => {
  const syntaxHighlighterEl = useRef(null);

  useEffect(() => {
    if (syntaxHighlighterEl.current) {
      syntaxHighlighterEl.current.scrollTop = 0;
      syntaxHighlighterEl.current.scrollLeft = 0;
    }
  }, [children]);

  return <pre {...restProps} ref={syntaxHighlighterEl}>{children}</pre>;
};

CodePreTag.propTypes = {
  children: PropTypes.node
};

export default CodePreTag;
