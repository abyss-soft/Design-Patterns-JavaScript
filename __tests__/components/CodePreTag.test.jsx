import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import 'jest-styled-components';

import CodePreTag from '../../src/components/CodePreTag';

const shortCodeToRender = `
  function foo() {
    return 'bar';
  }
`;

const longCodeToRender = `
  function EquipmentPattern(name) {
    this.equipments = [];
    this.name = name;
  }

  EquipmentPattern.prototype.add = function(equipment) {
    this.equipments.push(equipment);
  };

  EquipmentPattern.prototype.getPrice = function() {
    return this.equipments
      .map(function(equipment) {
        return equipment.getPrice();
      })
      .reduce(function(a, b) {
        return a + b;
      });
  };

  function Equipment() {}

  Equipment.prototype.getPrice = function() {
    return this.price;
  };

  // -- leafs
  function FloppyDisk() {
    this.name = 'Floppy Disk';
    this.price = 70;
  }
  FloppyDisk.prototype = Object.create(Equipment.prototype);

  function HardDrive() {
    this.name = 'Hard Drive';
    this.price = 250;
  }
  HardDrive.prototype = Object.create(Equipment.prototype);

  function Memory() {
    this.name = '8gb memomry';
    this.price = 280;
  }
  Memory.prototype = Object.create(Equipment.prototype);

  module.exports = [EquipmentPattern, FloppyDisk, HardDrive, Memory];
`;

describe('<CodePreTag /> component', () => {
  it('renders children properly', () => {
    const tree = renderer
      .create(
        <CodePreTag>
          <code>{shortCodeToRender}</code>
        </CodePreTag>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('scrolls to position 0,0 on children change', () => {
    const container = mount(
      <CodePreTag style={{ height: 200 }}>
        <code>{longCodeToRender}</code>
      </CodePreTag>
    );
    const $container = container.getDOMNode();

    // Initially the scroll position must be at the top
    expect($container.scrollTop).toBe(0);

    // Test that the scroll position is updated correctly
    $container.scrollTop = 100;
    expect($container.scrollTop).toBe(100);

    // Test that the scroll position is reset after changing its children
    container.setProps({ children: shortCodeToRender });
    expect($container.scrollTop).toBe(0);
  });
});
