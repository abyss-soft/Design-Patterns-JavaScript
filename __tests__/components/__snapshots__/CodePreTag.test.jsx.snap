// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CodePreTag /> component renders children properly 1`] = `
<pre>
  <code>
    
  function foo() {
    return 'bar';
  }

  </code>
</pre>
`;
